### PR TITLE
feat(Book): add search functionality for books 

### PR DESCRIPTION
**Pull request by zaidoon1**
## Description of changes
Added back-end functionality to allow for searching of books. Three algorithms have been implemented:

- fuzzySearch: we get a match even when search term is not exactly the same (few letters off)
- keywordSearch: basic search where if the search term contains a keyword we get a match. 
- wildcardSearch: we get a match even if we have a partial match

Other changes:

- Some tests have been moved from BookControllerTest to BookServiceTest as they don't test the BookController but the BookService.
- Unit and Integration tests have been refactored by removing unnecessary code.

if you would like to test the end-points, send a GET request to the following end-points AFTER adding a book:

- http://localhost:8098/fuzzySearchName?keywords=whateverTextYouWant
- http://localhost:8098/keywordSearchName?keywords=whateverTextYouWant
- http://localhost:8098/wildcardSearchName?keywords=whateverTextYouWant

Please note that I will add the integration and unit tests in a separate pr as I don't want this pr to get too big.

## Screenshots (If Applicable)
N/A

## Checklist
- [ ] I have unit tested my code.
- [ ] I have integration tested my code.
- [X] All new and existing tests passed.
- [ ] I have [updated the diagrams](https://github.com/amazin-team/Amazin-online-bookstore/wiki/Updating-the-diagrams) ModelsUMLClassDiagram and EntityRelationshipDiagram (if necessary).

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -89,15 +89,14 @@
...
             <groupId>nz.net.ultraq.thymeleaf</groupId>
             <artifactId>thymeleaf-layout-dialect</artifactId>
             <version>2.3.0</version>
-=======
+        </dependency>
+        <dependency>
             <groupId>org.hibernate</groupId>
             <artifactId>hibernate-search-orm</artifactId>
             <version>5.11.1.Final</version>
->>>>>>> Stashed changes
```
Comment (MohamedZalat) [inline pom.xml]: How is there a commit that had this in it?

File: pom.xml [hunk 2]
```diff
@@ -89,9 +89,15 @@
...
         <dependency>
+<<<<<<< Updated upstream
             <groupId>nz.net.ultraq.thymeleaf</groupId>
             <artifactId>thymeleaf-layout-dialect</artifactId>
             <version>2.3.0</version>
+=======
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-search-orm</artifactId>
+            <version>5.11.1.Final</version>
+>>>>>>> Stashed changes
```
Comment (MohamedZalat) [inline pom.xml]: Remove the merge lines from the commit

File: pom.xml (hunk 1 above)
Comment (zaidoon1) [reply pom.xml]: This has been fixed a long time ago

File: pom.xml (hunk 2 above)
Comment (zaidoon1) [reply pom.xml]: This has been fixed a long time ago

File: pom.xml (hunk 1 above)
Comment (MohamedZalat) [reply pom.xml]: None of the commits should have this, do an interactive rebase and remove it, as all commits need to build.

File: pom.xml (hunk 2 above)
Comment (MohamedZalat) [reply pom.xml]: Go to this commit and remove those lines.

File: pom.xml (hunk 1 above)
Comment (zaidoon1) [reply pom.xml]: should be fixed now.

File: pom.xml (hunk 2 above)
Comment (zaidoon1) [reply pom.xml]: should be fixed now.

**Review by MohamedZalat — APPROVED** [review]

**Pull request merged**